assayist/processor/loose_artifact_analyzer.py

```diff
@@ -49,104 +55,124 @@ def run(self):
+                    # file path.
+                    for relative_filepath, response in zip(relative_filepath_batch, responses):
+                        is_rpm = relative_filepath.endswith('.rpm')
+                        # If Koji could not find it or there was some other error, log it
+                        # and continue. Response is either a dict if an error, or a list of
+                        # one element if found.
+                        if isinstance(response, dict):
+                            log.error(f'Error received from Koji looking up {relative_filepath}'
+                                      f' embedded in {archive} in build {build_id}. Koji error '
+                                      f'{response["faultString"]}')
+                            continue
+
+                        artifact_info = response[0]
+                        if not artifact_info:
+                            log.info(f'Cannot find build for {relative_filepath} embedded in '
+                                     f'{archive} in build {build_id}.')
+                            continue
+
+                        if not is_rpm:
+                            # listArchives returns a list where getRPM returns a hash directly
+                            artifact_info = artifact_info[0]
+
+                        artifact_build_id = artifact_info.get('build_id')
+                        if not artifact_build_id:
+                            log.error(f'Empty build found in Koji for {relative_filepath} '
+                                      f'embedded in {archive} in build {build_id}')
+                            continue
+
+                        log.info(f'Linking discovered embedded artifact {relative_filepath} '
+                                 f'embedded in {archive} in build {build_id}')
+                        artifact_build = content.Build.get_or_create({
+                            'id_': artifact_build_id,
+                            'type_': 'build' if is_rpm else artifact_info['btype'],  # TODO bug!
+                        })[0]
+
+                        if is_rpm:
+                            artifact = self.create_or_update_rpm_artifact_from_rpm_info(
+                                artifact_info)
+                        else:
+                            artifact = self.create_or_update_archive_artifact_from_archive_info(
+                                artifact_info)
+
+                        self.conditional_connect(artifact.build, artifact_build)
+                        if is_archive:
                             self.conditional_connect(original_artifact.embedded_artifacts, artifact)
                             self.claim_file(path_to_archive, relative_filepath)
+                        else:
+                            # Just remember it for now so that we can connect with the build
+                            # artifacts later.
+                            embedded_in_source.append(artifact)
```

Comment:
I'm a bit confused why the found artifacts embedded in source are not connected right away and rather added to this list that is then processed in the next iteration?

Reply:
Well we want to connect them with every build artifact, which right now we don't have a list of. But we're going to be iterating through them anyway in later instances of the loop, so the intent here is just to link the embedded source artifacts as we are processing each build artifact.